neuromation/api/object_storage.py

```diff
@@ -0,0 +1,401 @@
+                        child_rel_path,
+                        filter=filter,
+                        queue=queue,
+                    )
+                )
+            else:
+                await queue.put(
+                    StorageProgressFail(
+                        src / name,
+                        dst / name,
+                        f"Cannot upload {child}, not regular file/directory",
+                    )
+                )
+        await _run_concurrently(tasks)
+        await queue.put(StorageProgressLeaveDir(src, dst))
+
+
+def _format_bucket_uri(bucket_name: str, key: str = "") -> URL:
+    return URL.build(scheme="object", host=bucket_name, path="/" + key.lstrip("/"))
+
+
+def _obj_status_from_bucket(data: Dict[str, Any]) -> ObjStatus:
+    mtime = datetime.datetime.fromisoformat(data["creation_date"]).timestamp()
+    return ObjStatus(
+        path=data["name"],
+        type=FileStatusType.DIRECTORY,
+        size=0,
+        modification_time=int(mtime),
+        bucket_name=data["name"],
+    )
+
+
+def _obj_status_from_key(bucket_name: str, data: Dict[str, Any]) -> ObjStatus:
+    type_ = (
+        FileStatusType.DIRECTORY if data["key"].endswith("/") else FileStatusType.FILE
+    )
+    return ObjStatus(
+        path=data["key"],
+        type=type_,
+        size=int(data["size"]),
+        modification_time=int(data["last_modified"]),
+        bucket_name=bucket_name,
+    )
+
+
+def _obj_status_from_prefix(bucket_name: str, data: Dict[str, Any]) -> ObjStatus:
+    return ObjStatus(
+        path=data["prefix"],
+        type=FileStatusType.DIRECTORY,
+        size=0,
+        modification_time=0,
+        bucket_name=bucket_name,
+    )
+
+
+async def _run_progress(
+    queue: "asyncio.Queue[ProgressQueueItem]",
+    progress: Optional[AbstractFileProgress],
+    coro: Awaitable[None],
+) -> None:
+    async def wrapped() -> None:
+        try:
+            await coro
+        finally:
+            await queue.put(None)
+
+    loop = asyncio.get_event_loop()
+    task = loop.create_task(wrapped())
+    while True:
+        item = await queue.get()
+        if item is None:
+            break
+
+        if progress is None:
+            # Just ignore event if we don't have a listener
+            continue
+        if isinstance(item, StorageProgressStart):
+            progress.start(item)
+        elif isinstance(item, StorageProgressComplete):
+            progress.complete(item)
+        elif isinstance(item, StorageProgressStep):
+            progress.step(item)
+        else:
+            progress = cast(AbstractRecursiveFileProgress, progress)
+
+            if isinstance(item, StorageProgressEnterDir):
+                progress.enter(item)
+            elif isinstance(item, StorageProgressLeaveDir):
+                progress.leave(item)
+            elif isinstance(item, StorageProgressFail):
+                progress.fail(item)
+
+    await task
+
+
+async def calc_md5(path: Path) -> str:
```

Comment:
Note the MD5 checksum. It's quite different and in Object Storages is optional, but I did not expose an option yet. Do we need it? It's quite fast, but can be intensive for large files (8GB file computes ~16s on my PC. That's 16s without output for user)